internal/credential/rotate.go

```diff
@@ -161,19 +158,29 @@ func (c *Config) ValidateServiceAccountKey(
 	var validatePermissionsErr error
 	var validationCallbackErr error
 
-	for start := time.Now(); time.Since(start) < validateServiceAccountKeyTimeout; time.Sleep(200 * time.Millisecond) {
-		_, validatePermissionsErr = c.ValidateIamPermissions(ctx, permissions, opts...)
+	ticker := time.NewTicker(200 * time.Millisecond)
+	defer ticker.Stop()
 
-		if validateCredsCallback != nil {
-			validationCallbackErr = validateCredsCallback(c, opts...)
-		}
+	timeout := time.After(c.validateServiceAccountKeyTimeout)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-timeout:
+			return errors.Join(validatePermissionsErr, validationCallbackErr)
```

Comment:
Should we not have a special error in the case of a timeout? Maybe something like
```suggestion
			return fmt.Errorf("failed to validate service account key before timeout: %w", errors.Join(validatePermissionsErr, validationCallbackErr))
```